### PR TITLE
feat: add gitlab provider for bootstrapping

### PR DESCRIPTION
**Pull request by Skarlso**
## Description

Closes https://github.com/open-component-model/MPAS/issues/62.

Looks like there is still a bug. If the repository doesn't exist, it will fail immediately because it creates the repository uninitialised. This means there is no `main` branch it can clone. This bug appears to be in Flux, though. So, I'm not sure what to do about it.

## What type of PR is this? (check all applicable)

- [ ] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

- Related Issue # (issue)
- Closes # (issue)
- Fixes # (issue)
> Remove if not applicable

## Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help
- [ ] Separate ticket for tests # (issue/pr)

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Added e2e-tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help
- [ ] Separate ticket for e2e-tests

## Added to documentation?

- [ ] 📜 README.md
- [ ] 🙅 no documentation needed

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Skarlso** [comment]
Test output:

```
➜  mpas git:(gitlab) ✗ ./bin/mpas bootstrap gitlab --owner skarlso --repository=mpas-test-project --personal
Running mpas bootstrap ...
 ✓   Preparing Management repository mpas-test-project
 ✓   Fetching bootstrap component from ghcr.io/open-component-model/mpas-bootstrap-component
 ✓   Installing flux with version v2.1.2
 ✓   Installing cert-manager with version v1.13.2
 ✓   Reconciling bootstrap components
 ✓   Waiting for cert-manager to be available
 ✓   Generating external-secrets-operator manifest with version v0.9.9
 ✓   Generating git-controller manifest with version v0.11.0
 ✓   Generating mpas-product-controller manifest with version v0.10.0
 ✓   Generating mpas-project-controller manifest with version v0.5.0
 ✓   Generating ocm-controller manifest with version v0.18.1
 ✓   Generating replication-controller manifest with version v0.12.2
 ✓   Generating certificate manifests
 ✓   Reconciling component manifests
 ✓   Waiting for components to be ready

Bootstrap completed successfully!
```

Screenshot 2023-12-11 at 18 28 31

**Comment by Skarlso** [comment]
Uh.

```
ERROR    Reconciler error    {"controller": "productdeployment", "controllerGroup": "mpas.ocm.software", "controllerKind": "ProductDeployment", "ProductDeployment": {"name":"podinfo","namespace":"mpas-test-001"}, "namespace": "mpas-test-001", "name": "podinfo", "reconcileID": "9813b386-8e70-45f1-8a29-706794ea3b30", "error": "failed to validate config: no config found"}                                                                                              
```

**Comment by Skarlso** [comment]
The GitRepositroy

```
Conditions:
    Last Transition Time:  2023-12-11T18:36:07Z
    Message:               building artifact
    Observed Generation:   1
    Reason:                ProgressingWithRetry
    Status:                True
    Type:                  Reconciling
    Last Transition Time:  2023-12-11T18:36:07Z
    Message:               failed to configure authentication options: failed to get secret 'mpas-test-001/gitea-credentials': secrets "gitea-credentials" not found
    Observed Generation:   1
    Reason:                AuthenticationFailed
    Status:                False
    Type:                  Ready
    Last Transition Time:  2023-12-11T18:34:32Z
    Message:               failed to configure authentication options: failed to get secret 'mpas-test-001/gitea-credentials': secrets "gitea-credentials" not found
    Observed Generation:   1
    Reason:                AuthenticationFailed
    Status:                True
    Type:                  FetchFailed
  Observed Generation:     -1
Events:
  Type     Reason                Age                  From               Message
  ----     ------                ----                 ----               -------
  Warning  AuthenticationFailed  77s (x7 over 2m52s)  source-controller  failed to configure authentication options: failed to get secret 'mpas-test-001/gitea-credentials': secrets "gitea-credentials" not found

```

What in tarnasion?

**Comment by Skarlso** [comment]
The above error is fixed by git-controller update. :)

File: internal/bootstrap/provider/provider.go [hunk 1]
```diff
@@ -83,6 +85,16 @@ func giteaProviderFunc(opts ProviderOptions) (gitprovider.Client, error) {
 	return client, err
 }
 
+func gitlabProviderFunc(opts ProviderOptions) (gitprovider.Client, error) {
+	o := makeProviderOpts(opts)
+	// TODO: Put that into an option somewhere.
```
Comment (phoban01) [inline internal/bootstrap/provider/provider.go:90]: Put **what** into an option somewhere?

**Review by phoban01 — APPROVED** [review]

**Comment by phoban01** [comment]
Looks good, few small issues.

File: cmd/mpas/bootstrap/bootstrap_gitlab.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+// Execute executes the command and returns an error if one occurred.
+func (b *GitlabCmd) Execute(ctx context.Context, cfg *config.MpasConfig) error {
+	t, err := time.ParseDuration(cfg.Timeout)
+	if err != nil {
+		return err
+	}
+	ctx, cancel := context.WithTimeout(ctx, t)
+	defer cancel()
+
+	//if b.Hostname == "" {
```
[the remark this answers is not in the document]
Comment (Skarlso) [reply cmd/mpas/bootstrap/bootstrap_gitlab.go:67]: Removed. Ops.

File: internal/bootstrap/provider/provider.go (hunk 1 above)
Comment (Skarlso) [reply internal/bootstrap/provider/provider.go:90]: ah right, sorry. This thing: `"oauth2"`. I need to do something about it. But forgot. :D

**Pull request merged**